### ISSUE TITLE
[casl-react] usePermission hook proposal

### ISSUE THREAD

**Issue by dacevedo12** [stalniy/casl#291]
> In this case, you need to create a new Ability instance when you want to update user permissions (don't use update, it won't trigger re-rendering in this case) or you need to force re-render the whole app.

I first encountered the issue and then read about it in the docs. After reading the `<Can>` component code i figured that we might just extract the on('update') behaviour to a hook, export it and also reuse it there.

What do you guys think? Any considerations (maybe performance/rerendering) ?

**requirements**:
- [x] create `useAbility` hook that accepts `AbilityContext` and returns bound `Ability` instance
- [x] hook should force component to update if ability rules are changed 
- [x] add tests for hook
- [x] add docs about hook

**Comment by stalniy** [comment]
`usePermission` hook is an interesting idea!

What API do you think is the best? Something like this?

```js
const canUpdate = usePermission(‘update’, ‘Todo’)
```

Or do you have something better in mind?

**Comment by dacevedo12** [comment]
The idea i had in mind was aiming to solve the problem that a bare useContext has. It wont trigger a rerender and i have use cases where i must go for the imperative way. 

Do you think it is possible to subscribe to all permissions?

`const { can } = usePermission(context);`

**Comment by stalniy** [comment]
I see. Think makes much more sense to me :) But I think we can even do this:

```js
import { useAbility } from '@casl/react/hooks'

const { can, cannot } = useAbility(); // the context is optional
```

I used `useAbility` instead of `userPermission` because it's one of the CASL terminology. This hook should be define in a separate submodule (`@casl/react/hooks`) in order to not introduce breaking changes. 

Can you provide a PR that do this?

**Comment by dacevedo12** [comment]
Yeah sure, i'll try it on my implementation first and then open a PR. 

Just a doubt tho, how would it be possible for the context to be optional? The Can component needs to be bound, woudnt we need either that or passing it as argument there?

**Comment by stalniy** [comment]
We can create default one in hooks submodule. 

```js
export const AbilityContext = createContext();

export function useAbility(context = AbilityContext){
  // implementation 
}
```

Or do you think it’s useless?

**Comment by dacevedo12** [comment]
That would be a gotcha if someone needs both `<Can>` and useAbility and doesn't specify the context. The 2 different contexts would not have the same abilities.

I think we can do an initial implementation with context being a mandatory argument and if someone finds having a default useful that can be added later

**Comment by stalniy** [comment]
OK, let's start with mandatory option

**Comment by stalniy** [comment]
released in `@casl/react@2.1.0`. You can import hook as this:

```js
import { useAbility } from '@casl/react'
```
